### PR TITLE
Moving metrics related exec calls from zedagent to volumemgr.

### PR DESCRIPTION
**Pull request by adarsh-zededa**
Issue:
We did a bunch of system calls every 15s to calculate diskmetrics which was to be reported to the controller. 
This caused 2 issued:
1. That many system calls, every 15s used up a lot a memory.
2. Some disk metrics doesn't change that often to be recalculated.

Solution:
Moved all disk related system call operations from zedagent to the volumemgr. Now volumemgr publishes a diskMetricStatus (for disks and mountPaths) and appDiskMetricStatus (for volumeStatuses) for every 5 min. Other than that, every time we create or update a volumeStatus, appDiskMetricStatus will be created/updated just for that volume and will be published.
On the zedagent's front, it will still send disk metrics to the controller every 15s, but instead of calculating everything, it will just take the latest list of diskMetricStatus and appDiskMetricStatus that sent by volumemgr and report that.

Note: Introduced a new subservice in volumemgr called `volumemgrmetrics`.

Signed-off-by: adarsh-zededa <adarsh@zededa.com>

### CODE REVIEW

File: pkg/pillar/types/diskmetrics.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+// LogDelete :
+func (status AppDiskMetricStatus) LogDelete() {
+	logObject := base.EnsureLogObject(nil, base.AppDiskMetricStatusType, status.DiskPath, nilUUID, status.LogKey())
+	logObject.CloneAndAddField("diskid", status.DiskID).
+		AddField("diskpath", status.DiskPath).
+		AddField("provisionedbytes-int64", status.ProvisionedBytes).
+		AddField("disktype", status.DiskType).
+		AddField("dirty", status.Dirty).
+		Infof("AppDiskMetric status delete")
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:173]: Please change to Metricf

File: pkg/pillar/types/diskmetrics.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	if _, ok := old.(AppDiskMetricStatus); !ok {
+		logObject.Clone().Fatalf("LogModify: Old object interface passed is not of AppDiskMetricStatus type")
+	}
+	logObject.CloneAndAddField("diskid", status.DiskID).
+		AddField("diskpath", status.DiskPath).
+		AddField("old-userbytes-int64", status.UsedBytes).
+		AddField("provisionedbytes-int64", status.ProvisionedBytes).
+		AddField("disktype", status.DiskType).
+		AddField("dirty", status.Dirty).
+		Infof("AppDiskMetric status modify")
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:162]: Please change to Metricf

File: pkg/pillar/types/diskmetrics.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	if logObject == nil {
+		return
+	}
+	logObject.CloneAndAddField("diskid", status.DiskID).
+		AddField("diskpath", status.DiskPath).
+		AddField("userbytes-int64", status.UsedBytes).
+		AddField("provisionedbytes-int64", status.ProvisionedBytes).
+		AddField("disktype", status.DiskType).
+		AddField("dirty", status.Dirty).
+		Infof("AppDiskMetric status create")
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:146]: Please change to Metricf

File: pkg/pillar/types/diskmetrics.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+// LogDelete :
+func (status DiskMetricStatus) LogDelete() {
+	logObject := base.EnsureLogObject(nil, base.DiskMetricStatusType, status.DiskPath, nilUUID, status.LogKey())
+	logObject.CloneAndAddField("diskid", status.DiskID).
+		AddField("diskpath", status.DiskPath).
+		AddField("mountpath", status.MountPath).
+		AddField("totalbytes-int64", status.TotalBytes).
+		AddField("userbytes-int64", status.UsedBytes).
+		AddField("freebytes-int64", status.FreeBytes).
+		Infof("DiskMetric status delete")
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:109]: Please change to Metricf

File: pkg/pillar/types/diskmetrics.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+		AddField("diskpath", status.DiskPath).
+		AddField("mountpath", status.MountPath).
+		AddField("old-readbytes-int64", status.ReadBytes).
+		AddField("old-writebytes-int64", status.WriteBytes).
+		AddField("old-readcount-int64", status.ReadCount).
+		AddField("old-writecount-int64", status.WriteCount).
+		AddField("totalbytes-int64", status.TotalBytes).
+		AddField("old-userbytes-int64", status.UsedBytes).
+		AddField("old-freebytes-int64", status.FreeBytes).
+		Infof("DiskMetric status modify")
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:97]: Please change to Metricf

File: pkg/pillar/types/diskmetrics.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+		AddField("diskpath", status.DiskPath).
+		AddField("mountpath", status.MountPath).
+		AddField("readbytes-int64", status.ReadBytes).
+		AddField("writebytes-int64", status.WriteBytes).
+		AddField("readcount-int64", status.ReadCount).
+		AddField("writecount-int64", status.WriteCount).
+		AddField("totalbytes-int64", status.TotalBytes).
+		AddField("userbytes-int64", status.UsedBytes).
+		AddField("freebytes-int64", status.FreeBytes).
+		Infof("DiskMetric status create")
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:77]: Please change to Metricf

File: pkg/pillar/types/diskmetrics.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+// If so how do we tell them apart
+var AppPersistPaths = []string{
+	VolumeEncryptedDirName,
+	VolumeClearDirName,
+	SealedDirName + "/downloader",
+	SealedDirName + "/verifier",
+}
+
+//DiskMetricStatus holds metrics data per disk
+type DiskMetricStatus struct {
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:43]: Please name this DiskMetric; Status makes sense as a response to Config but this is not that.

File: pkg/pillar/cmd/volumemgr/handlediskmetrics.go [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+	log.Debugf("lookupAppDiskMetricStatus(%s) Done", key)
+	return &status
+}
+
+//diskMetricsTimerTask calculates and publishes disk metrics periodically
+func diskMetricsTimerTask(ctx *volumemgrContext) {
+	log.Infoln("starting report diskMetricsTimerTask timer task")
+	createOrUpdateDiskMetrics(ctx)
+
+	diskMetricInterval := time.Duration(300) * time.Second
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/handlediskmetrics.go:118]: How about making this a ConfigItem so it can be set from the controller?

File: pkg/pillar/cmd/zedagent/handlemetrics.go [hunk 9]
```diff
@@ -21,44 +22,111 @@ import (
...
 
-// Application-related files live here
-// XXX do we need to exclude the ContentTrees used for eve image update?
-// If so how do we tell them apart
-var appPersistPaths = []string{
-	types.VolumeEncryptedDirName,
-	types.VolumeClearDirName,
-	types.SealedDirName + "/downloader",
-	types.SealedDirName + "/verifier",
+func lookupDiskMetricStatus(ctx *zedagentContext, diskPath string) *types.DiskMetricStatus {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlemetrics.go:55]: This function doesn't appear to be used.

Do we need to lookup on either diskPath and mountPath? If we can stick to one we can make that the Key and avoid doing a GetAll below.

**Review by eriknordmark — COMMENTED** [review]
LGTM. A few minor issues as separate comments.

File: pkg/pillar/cmd/zedagent/handlemetrics.go (hunk 9 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/zedagent/handlemetrics.go]: Changed.

File: pkg/pillar/cmd/volumemgr/handlediskmetrics.go (hunk 8 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/volumemgr/handlediskmetrics.go]: Done.

File: pkg/pillar/types/diskmetrics.go (hunk 7 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: Changed.

File: pkg/pillar/types/diskmetrics.go (hunk 6 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: changed.

File: pkg/pillar/types/diskmetrics.go (hunk 5 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: changed.

File: pkg/pillar/types/diskmetrics.go (hunk 4 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: changed.

File: pkg/pillar/types/diskmetrics.go (hunk 3 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: changed.

File: pkg/pillar/types/diskmetrics.go (hunk 2 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: changed.

File: pkg/pillar/types/diskmetrics.go (hunk 1 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: changed.

File: pkg/pillar/utils/globalutils.go [hunk 10]
```diff
@@ -6,12 +6,11 @@
 package utils
 
 import (
-	"os"
-
 	"github.com/lf-edge/eve/pkg/pillar/base"
 	"github.com/lf-edge/eve/pkg/pillar/pubsub"
 	pubsublegacy "github.com/lf-edge/eve/pkg/pillar/pubsub/legacy"
 	"github.com/lf-edge/eve/pkg/pillar/types"
+	"os"
```
Comment (eriknordmark) [inline pkg/pillar/utils/globalutils.go:13]: We are gradually moving in the other directioj to have the standard packages like "os" up front followed by the github.com etc packages. So please revert this change.

File: pkg/pillar/cmd/zedagent/handlemetrics.go [hunk 11]
```diff
@@ -1287,3 +1307,10 @@ func protoEncodeNetworkInstanceMetricProto(status types.NetworkInstanceMetrics)
 
 	return metric
 }
+
+// pathToKey converts /path/to/dir to path-to-dir.
+func pathToKey(path string) string {
+	path = strings.TrimPrefix(path, "/")
+	key := strings.ReplaceAll(path, "/", "-")
+	return key
+}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlemetrics.go:1316]: You already have this in types package; better to export that function instead of duplicating it.

File: pkg/pillar/cmd/volumemgr/handlediskmetrics.go [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+	return &status
+}
+
+//diskMetricsTimerTask calculates and publishes disk metrics periodically
+func diskMetricsTimerTask(ctx *volumemgrContext) {
+	log.Infoln("starting report diskMetricsTimerTask timer task")
+	createOrUpdateDiskMetrics(ctx)
+
+	diskMetricInterval := time.Duration(ctx.globalConfig.GlobalValueInt(types.DiskScanMetricInterval)) * time.Second
+	max := float64(diskMetricInterval)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/handlediskmetrics.go:104]: This doesn't seem to react to changes to the configuration.
In handlemetrics.go we have logic to do that for  updateMetricsTimer plus code in parseConfigItem to call it.
That makes sense for this timer as well.

File: pkg/pillar/types/diskmetrics.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+//DiskMetric holds metrics data per disk
+type DiskMetric struct {
+	DiskPath   string
+	ReadBytes  uint64
+	WriteBytes uint64
+	ReadCount  uint64
+	WriteCount uint64
+	TotalBytes uint64
+	UsedBytes  uint64
+	FreeBytes  uint64
```
Comment (eriknordmark) [inline pkg/pillar/types/diskmetrics.go:52]: Please add comment to specify the units for these; bytes vs. MBytes. Note that the protobuf fields are in MBytes as specified in the comment in the proto file.

File: pkg/pillar/cmd/zedagent/handlemetrics.go [hunk 14]
```diff
@@ -383,62 +417,40 @@ func publishMetrics(ctx *zedagentContext, iteration int) {
...
-		log.Debugf("Path %s total %d used %d free %d",
-			path, u.Total, u.Used, u.Free)
-		metric := metrics.DiskMetric{MountPath: path,
-			Total: utils.RoundToMbytes(u.Total),
-			Used:  utils.RoundToMbytes(u.Used),
-			Free:  utils.RoundToMbytes(u.Free),
+		metric := metrics.DiskMetric{
+			Disk:       diskPath,
+			MountPath:  mountPath,
+			ReadBytes:  diskMetric.ReadBytes,
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlemetrics.go:431]: Somewhere (here or in zedagent) this needs to be converted/rounded to Mbytes since that is what should be iin the protobuf message.

File: pkg/pillar/utils/globalutils.go (hunk 10 above)
Comment (adarsh-zededa) [reply pkg/pillar/utils/globalutils.go]: Done.

File: pkg/pillar/cmd/zedagent/handlemetrics.go (hunk 11 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/zedagent/handlemetrics.go]: Done.

File: pkg/pillar/cmd/volumemgr/handlediskmetrics.go (hunk 12 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/volumemgr/handlediskmetrics.go:103]: Changed.

File: pkg/pillar/types/diskmetrics.go (hunk 13 above)
Comment (adarsh-zededa) [reply pkg/pillar/types/diskmetrics.go]: Added.

File: pkg/pillar/cmd/zedagent/handlemetrics.go (hunk 14 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/zedagent/handlemetrics.go]: Changed.
